### PR TITLE
Turshi instruction logic

### PR DESCRIPTION
**Pull request by querolita**
This PR provides a functioning version of the Cairo instruction logics gate constraints. It still doesn't have the wiring between I/O registers and public input, but it should be easily addable. This PR does not contain the memory logics, which has been moved to a different PR for readability and ease of review. In particular, some parts of the memory logics are subject to changes for lookup tables, so that part is less stable. It needs about 4 rows per Cairo instruction in a 15-wire kimchi. The delta between this PR and the one for the Cairo runner are the following: 
- Added one field to `CairoProgram` that gives the list of instructions of the execution.
- `turshi.rs`: includes 
  - 4 Cairo gates definitions; 
  -  a witness generator from an execution of the runner; 
  -  a gadget generation from a witness, 
  -  constraints for the operands, operations, registers, instruction checks, registers updates
  - a checking function that replicates the constraints expressions in field format
  - a verification function that evaluates the constraints directly
  - one simple check using the same program as the runner's tests
- some changes needed to be added to prover, expr, verifier, gate, wires,  index, for it to work. But further modifications will have to be carried out

These are the following types of gates (more or less, in gadget order). 
- `CairoClaim`: takes care of the initial and final registers claim
- `CairoInstruction`: takes care of the logics of the Cairo instruction.
- `CairoFlags`: performs registers update checks using the values of the flags
- `CairoTransition`: performs the check of a correct transition between consecutive instructions
- `Zero`: does nothing, it's just a row that holds 3 values. It would be great to merge this one somewhere else, but it doesn't seem to fit. So it is a `Zero` gate for now (used to be `CairoAuxiliary`, but I think it can be `Zero` like in ChaCha).

Some remarks/TODOs for future versions:
- I have to write the wiring of the Cairo rows. In the comments I wrote some sort of table of the current layout and it explains what columns should be copied and from where. But this is still pending.
- I want to create full tests that involve both the prover and the verifier, not only these "gates checks"
- (Issue [282](https://github.com/o1-labs/proof-systems/issues/282))Think about a different number of columns and how much it would benefit Cairo and other complex gates. In this approach, if `n` is the number of instructions, the full gadget for the program takes `4n-1` rows.
- (Issue [294](https://github.com/o1-labs/proof-systems/issues/294))The runner still needs to address the hints independently: meaning, run the program `backwards` and figure out the intermediate 2 or 3 values of the memory that need to be set for this nondeterministic feature of the memory

### CODE REVIEW

File: cairo/src/runner.rs
Comment (jspada) [inline cairo/src/runner.rs:527]: Fix comment.  Should be "the initial values of the ..."

**Comment by jspada** [comment]
Awesome work-- Looks great!

**Review by jspada — APPROVED** [review]
Approving,  but please see comments/feedback and wait for others to review also before merging.

File: cairo/src/word.rs
Comment (querolita) [reply cairo/src/word.rs:32]: Sure!

File: cairo/src/helper.rs
Comment (querolita) [reply cairo/src/helper.rs:11]: :)

File: cairo/src/word.rs
Comment (querolita) [reply cairo/src/word.rs:232]: it's already there, isn't it? this one:
`// opcode = 4*fOPC_AEQ + 2*fOPC_RET + fOPC_CALL`

File: kimchi/src/circuits/polynomials/chacha.rs
Comment (querolita) [reply kimchi/src/circuits/polynomials/chacha.rs:595]: sure, like above

File: kimchi/src/tests/turshi.rs
Comment (querolita) [reply kimchi/src/tests/turshi.rs:143]: all of the `assert_eq` are handled within `ensure_cairo_gate` and if there's an error, `res` will be an error as well

File: cairo/src/helper.rs
Comment (mimoo) [reply cairo/src/helper.rs:11]: `of_lsb` is really an ocaml thing. in Rust just I've seen `lsb`, or `to_lsb` (or `into_lsb` if you're consuming `self` and `self` is not `Copy`)

File: cairo/src/helper.rs
Comment (mimoo) [inline cairo/src/helper.rs:59]: note that this function exists here: https://github.com/o1-labs/proof-systems/blob/master/utils/src/field_helpers.rs#L90

File: cairo/src/runner.rs
Comment (mimoo) [inline cairo/src/runner.rs:10]: where is this used?

File: kimchi/src/circuits/constraints.rs
Comment (mimoo) [inline kimchi/src/circuits/constraints.rs:155]: since Cairo is optional, I think these should live under some `Option<Cairo>` field

File: kimchi/src/circuits/constraints.rs
Comment (mimoo) [inline kimchi/src/circuits/constraints.rs:697]: move the cairo initialization to its own function which you can call here :o

File: kimchi/src/linearization.rs
Comment (mimoo) [inline kimchi/src/linearization.rs:36]: can we do it like the lookup stuff? This way if we don't use cairo the alpha ranges are not affected

File: kimchi/src/proof.rs
Comment (mimoo) [inline kimchi/src/proof.rs:40]: import turshi so you can write `turshi::CIRCUIT_GATE_COUNT`

Comment (mimoo) [reply kimchi/src/proof.rs:40]: like lookup this should be an `Option`

**Review by mimoo — CHANGES_REQUESTED** [review]
Could you separate the code that touches the proof system in a separate PR? I think they should be reviewed separately :o

File: kimchi/src/tests/turshi.rs
Comment (querolita) [reply kimchi/src/tests/turshi.rs:75]: Every time you write to memory, the length gets automatically extended. Meaning, before line 72 the length was 21 (from 0 to 20) and then in line 75 the length of memory is already 24. the argument `24` in the creator of the program corresponds to the initial allocation pointer. This means, that I think you're right and this argument could be "omitted": the initial allocation pointer should be the first empty address of the memory, so it must be 24 in this case.

File: cairo/src/runner.rs
Comment (querolita) [reply cairo/src/runner.rs:527]: thanks!

File: cairo/src/helper.rs
Comment (querolita) [reply cairo/src/helper.rs:59]: You're right, good catch, I fixed this in the [i128 field helper](https://github.com/o1-labs/proof-systems/pull/493/commits/7dca96d7b3de965450947141c41ff95412def4e8) commit. Actually, I suggested that in that field_helper we could actually perform the casting using `u32` instead of `u64` to convert `i32` to fields.

File: cairo/src/runner.rs
Comment (querolita) [reply cairo/src/runner.rs:10]: Actually, this is being used only in the memory constraints, but I must have missed to remove this one from this PR.

File: kimchi/src/circuits/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/constraints.rs:155]: amazing 😃

File: utils/src/field_helpers.rs
Comment (mimoo) [inline utils/src/field_helpers.rs:102]: this still looks dangerous to me :o casting `i128` to `u128`

Comment (jspada) [reply utils/src/field_helpers.rs:102]: The signed version should be about half the size of the unsigned version, but to be safe maybe use the `From` or `TryFrom` trait?

Example `-F::from(u128::from(i))`

Or guard with `std::u128::MAX`

File: cairo/src/word.rs
Comment (jspada) [reply cairo/src/word.rs:232]: I just mean to make the code's form match the comment.

File: kimchi/src/tests/turshi.rs
Comment (jspada) [reply kimchi/src/tests/turshi.rs:75]: Hmm... yeah, this is interesting.  It might be cool if the memory thing was "smart" and you can just fill it up with the initial memory and the `CairoProgram::new(&mut mem)` was smart enough to figure out what to tuse for `pc` and `ap`.  Maybe a thing for future though and not this PR :-). I was just concerned if there could be a bad offset there that could impact the test case

File: utils/src/field_helpers.rs
Comment (mimoo) [reply utils/src/field_helpers.rs:102]: there are no From implementations usually when the conversion can be dangerous

Comment (mimoo) [reply utils/src/field_helpers.rs:102]: you could use TryFrom and return an error (or panic if you know that this will never be given as a malicious input)

File: kimchi/src/linearization.rs
Comment (querolita) [reply kimchi/src/linearization.rs:36]: Good point, I will give it a thought

File: kimchi/src/proof.rs
Comment (querolita) [reply kimchi/src/proof.rs:40]: I see, like above with the `cairo` and `cairo8`

File: utils/src/field_helpers.rs
Comment (querolita) [reply utils/src/field_helpers.rs:102]: Oh this chat is very enlightening... Nonetheless, this edge-case you're talking about is not the one we have here right? There you're saying:
`    let a = -i128::MAX;
    let b = a as u128;`
And then `a = -170141183460469231731687303715884105727`
and `b=170141183460469231731687303715884105729` 

But the conditional in the helper is instead doing this:
`    let a = -i128::MAX;
    let b = -a as u128;`
And then `a = -170141183460469231731687303715884105727`
and `b=170141183460469231731687303715884105727` 

Let's just use some smaller data so it is easier to imagine what's going on:
`let a = -i8::MAX;`
`let b = a as u8;`

In this case, `a = 0b1000_0001 as i8 => -127` and `b = 0b1000_0001 as u8 =>129`

But if we perform `b = -a as u8` , first, it negates obtaining +127 and then 127 is expressed as u8 with `0b0111_1111`.

Nonetheless, I think what you meant instead is the edge-case where `a=-i8::MIN`, meaning -128. In that case, performing `b=-a as u8` effectively tries to perform -(-128) which does not fit inside a `u8`, even if later on +128 fits inside `u8` just fine. 

So you're right! There's that issue when you try to convert the minimum value of the integer range. I am using this helper only to convert a bytecode program to field elements in the tests, without having the user write directly field elements, and I was not using `i128::MIN` for that. But if a helper is meant to work in any scenario, then I agree that this must be fixed or avoided somehow. Nonetheless, this should be upgraded so that indeed one could write field elements directly, as in the playground. I will fix this in the next commit of the pr.

File: kimchi/src/circuits/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/constraints.rs]: I am not sure I get it. None of the other gates have an independent function for this, do they?

**Comment by querolita** [comment]
Update: I have addressed @mimoo 's comments regarding the optional feature of the cairo gates, so I wrapped these fields in optional types. This way, powers of alpha are not altered if the circuit has no cairo gates. For this, I followed the same strategy as chacha, equipping the `constraints_expr` and `expr_linearization` with a boolean type cairo. OCaml types are updated to be optional as well, and the alphas have been updated on the specs.

This way, cairo gates have no effect on the rest of the proof system, but if you still think it is worth moving these things to another PR, that can be done.

File: utils/src/field_helpers.rs
Comment (mimoo) [reply utils/src/field_helpers.rs]: oh you're right, I actually didn't think of the edge case properly. So indeed this causes an integer overflow:

```rust
let a = i128::MIN;
let b = -a as u128;
```

 https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5770463ba987cd28eb7b0647fa397931

File: cairo/src/helper.rs
Comment (mimoo) [inline cairo/src/helper.rs:14]: please let's avoid `of_` which is an OCaml thing :D

File: book/src/specs/kimchi.md
Comment (mimoo) [inline book/src/specs/kimchi.md:1300]: there's still this path that could be imported :o

**Comment by mimoo** [comment]
> if you still think it is worth moving these things to another PR, that can be done.

yeah sorry can you do it ^^? I'd really like to be able to review the parts that affect kimchi in isolation so we make sure that it won't affect the hardfork

File: cairo/src/helper.rs
Comment (querolita) [reply cairo/src/helper.rs:14]: Ah ok! Both sounded good to me, I didn't know there was a standard

File: book/src/specs/kimchi.md
Comment (querolita) [reply book/src/specs/kimchi.md:1300]: oh there! your eyes are well trained 👀

**Comment by querolita** [comment]
> yeah sorry can you do it ^^? I'd really like to be able to review the parts that affect kimchi in isolation so we make sure that it won't affect the hardfork

sgtm, I'll do that then :)

File: utils/src/field_helpers.rs
Comment (mimoo) [reply utils/src/field_helpers.rs]: makes me think that the function above (`i32_to_field`) could also be simplified as a one-liner `F::from(i.abs() as u64)`

File: cairo/src/helper.rs
Comment (mimoo) [reply cairo/src/helper.rs:14]: yeah I think I replied on the comment there. But in rust we use `from` and `into` and `to` generally. (Usually `into` is used when `self` is consumed and does not implement `Copy`)

File: kimchi/src/circuits/constraints.rs
Comment (mimoo) [reply kimchi/src/circuits/constraints.rs]: if you look at lookup it does

Comment (mimoo) [reply kimchi/src/circuits/constraints.rs]: see https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/constraints.rs#L345

File: cairo/src/helper.rs
Comment (querolita) [reply cairo/src/helper.rs:14]: oh I was not aware of the `into` distinction for consuming the `self`! nice lesson

File: kimchi/src/circuits/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/constraints.rs]: Oh I see, for lookup yeah... I was looking at chacha for instance. But fine, I can make it separately.

File: cairo/src/word.rs
Comment (querolita) [reply cairo/src/word.rs:221]: oh well you mean to write the 4 directly and so... Yeah well this was to avoid some multiplications here and there

File: kimchi/src/verifier.rs
Comment (mimoo) [inline kimchi/src/verifier.rs:646]: there's also `unimplemented!()` as well.

**Review by mimoo — APPROVED** [review]
Impressive work :)

File: kimchi/src/alphas.rs
Comment (querolita) [reply kimchi/src/alphas.rs:335]: absolutely right, that was a print I wrote for debugging, and seems like I forgot to remove it 😵‍💫

File: kimchi/src/circuits/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/constraints.rs]: Oh you're right that's way cleaner! I will modify this in the `turshi_pfsys` branch 😃

File: kimchi/src/verifier.rs
Comment (querolita) [reply kimchi/src/verifier.rs:646]: indeed 😅

File: kimchi/src/circuits/polynomials/turshi.rs
Comment (querolita) [reply kimchi/src/circuits/polynomials/turshi.rs:293]: oh really good point! I was only foreseeing the use case where you only had a huge cairo gadget corresponding to a program, so it didn't have to interact with others. Nonetheless, adding next row is super simple to add and can be useful in the future! 👍🏻

File: kimchi/src/circuits/polynomials/turshi.rs
Comment (querolita) [reply kimchi/src/circuits/polynomials/turshi.rs:327]: Oh that's fine, but only for `ensure_`,  since `verify_` is the one called from `gates.rs` for all other gates. Nonetheless, I can also create a `mod witness` for all things related to creating the witness, again to put some order inside `turshi.rs`, which is too long otherwise.

File: kimchi/src/circuits/expr.rs
Comment (querolita) [reply kimchi/src/circuits/expr.rs:417]: oh cute, didn't hear about this one

File: cairo/src/runner.rs
Comment (querolita) [reply cairo/src/runner.rs:102]: Good point, yes, `CairoContext` is copiable as well, and `CairoWord`

File: cairo/src/runner.rs
Comment (querolita) [reply cairo/src/runner.rs:662]: Oh! I was afraid that it wouldn't be able to convert from `i64` directly, it seems like it is indeed implemented. Thanks for the tip!

**Comment by querolita** [comment]
Merging Cairo instruction-related gates after addressing your comments (upcoming PRs about the proof system and the memory model)

File: poly-commitment/src/commitment.rs
Comment (mimoo) [reply poly-commitment/src/commitment.rs]: yeah I started getting this error on my own PRs too, this is weird :D

File: kimchi/src/circuits/constraints.rs
Comment (querolita) [reply kimchi/src/circuits/constraints.rs]: Yeah, combined with some tricks this helped reduce the function by 10 loc

**Pull request merged**
